File: compiler/src/queries/definitions.ts
import { Resolved } from "../visit";
import { query } from "./query";

export const unresolved = query(function* (node) {
    const resolved = node.facts.get(Resolved);
    if (resolved == null) {
        return;
    }

    const [name, definitions] = resolved;
    if (definitions.length === 0) {
        yield { name };
    }
});

export const ambigious = query(function* (node) {
    const resolved = node.facts.get(Resolved);
    if (resolved == null) {
        return;
    }

    const [name, definitions] = resolved;
    if (definitions.length > 1) {
        yield { name, definitions };
    }
});

File: compiler/queries/definitions.go
package queries

import (
	"wipple/database"
	"wipple/visit"
)

func Unresolved(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(name string)) {
	fact, ok := database.GetFact[visit.ResolvedFact](node)
	if !ok {
		return
	}

	if len(fact.Definitions) == 0 {
		f(fact.Name)
	}
}

func Ambiguous(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(name string)) {
	fact, ok := database.GetFact[visit.ResolvedFact](node)
	if !ok {
		return
	}

	if len(fact.Definitions) > 1 {
		f(fact.Name)
	}
}
